Harrison/agents rewrite

hwchase17 opened the pull request:
havent marked the class as deprecated yet, will likely want to do all in one go (with other classes)

hwchase17 committed:
add methods to deserialize prompts that were old

hwchase17 committed:
Merge branch 'master' of github.com:hwchase17/langchain

hwchase17 committed:
Merge branch 'master' of github.com:hwchase17/langchain

hwchase17 committed:
Merge branch 'master' of github.com:hwchase17/langchain

hwchase17 committed:
add conceptual agent docs

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
Merge branch 'master' of github.com:hwchase17/langchain

hwchase17 committed:
Merge branch 'master' of github.com:hwchase17/langchain

hwchase17 committed:
agents rewrite

baskaryan commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+
+
+def create_openai_functions_agent(
+    llm: BaseLanguageModel, tools: Sequence[BaseTool], prompt: ChatPromptTemplate
```
does prompt need to be typed as ChatPromptTemplate

baskaryan replied:
guessing it's an intentional choice not to make it optional?

hinthornw replied:
[Not design related] Since they're deprecating the functions API shouldn't we be pushing a "tools" agent?

baskaryan reviewed (commented):
if llm, tools, prompt are all we need for every agent maybe we can have a generic `create_agent(type, llm, tools, prompt)` method. don't need to do rn, just thinking aloud

baskaryan commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+            `include_chat_history` is set to True. It should be a list of LangChain
+            Message objects.
+    """
+    messages: List[Union[BaseMessagePromptTemplate, BaseMessage]]
+    messages = [SystemMessage(content=system_message)]
+    if include_chat_history:
+        messages.append(MessagesPlaceholder(variable_name="chat_history"))
+    messages.extend(
+        [
+            HumanMessagePromptTemplate.from_template("{input}"),
+            MessagesPlaceholder(variable_name="agent_scratchpad"),
+        ]
+    )
+    return ChatPromptTemplate(messages=messages)
```
nit: find ChatPromptTemplate.from_messages cleaner
```python
messages = [("system", system_message),]
if include_chat_history:
  messages.append(MessagesPlaceholder())
messages.extend([
  ("human", "{input}"),
  MessagesPlaceholder()
])
return ChatPromptTemplate.from_messages(messages)
```

jacoblee93 commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+    messages: List[Union[BaseMessagePromptTemplate, BaseMessage]]
+    messages = [SystemMessage(content=system_message)]
+    if include_chat_history:
+        messages.append(MessagesPlaceholder(variable_name="chat_history"))
```
I would really love to find a good way to standardize or reserve this name - always struck me as a bit arbitrary and inconsistent across modules

jacoblee93 replied:
Possibly as memory becomes more defined

baskaryan commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+    agent = (
+        RunnablePassthrough.assign(
+            agent_scratchpad=lambda x: format_to_openai_function_messages(
+                x["intermediate_steps"]
```
for trying to run agent manually (while experimenting) would be nice if "intermediate_steps" wasn't required (ie do `x.get('intermediate_steps', [])`)

baskaryan commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+        functions=[format_tool_to_openai_function(t) for t in tools]
+    )
+    agent = (
+        RunnablePassthrough.assign(
```
should we add a nice run_name

hinthornw commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+        - `input`: this is the user input
+        - `agent_scratchpad`: this is the history of steps the agent has taken so far.
+            This gets populated by the AgentExecutor.
+        - `chat_history`: This is optionally required as input, if
```
"optionally required" is confusing

jacoblee93 commented on libs/langchain/langchain/agents/openai_functions_agent/base.py:
```diff
@@ -226,3 +227,120 @@ def from_llm_and_tools(
+    return ChatPromptTemplate(messages=messages)
+
+
+def create_openai_functions_agent(
```
Worth renaming to `create_functions_agent` and having typing be some kind of `FunctionCallingChatModel` since you could do this with Anthropic/potentially Fireworks/Ollama?

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

hwchase17 committed:
stash

hwchase17 committed:
stash

hwchase17 committed:
Merge branch 'master' into harrison/agent-docs-concepts

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

hwchase17 committed:
Merge branch 'harrison/agent-docs-concepts' into harrison/agents-rewrite

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

nfcampos committed:
Add message chunks to stream_log output

hwchase17 committed:
cr

hwchase17 committed:
Merge branch 'harrison/agents-rewrite' of github.com:hwchase17/langchain into harrison/agents-rewrite

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

jacoblee93 commented on libs/langchain/langchain/agents/xml/base.py:
```diff
@@ -97,3 +102,70 @@ async def aplan(
+
+    prompt = prompt.partial(
+        tools=render_text_description(list(tools)),
+        tool_names=", ".join([t.name for t in tools]),
```
Don't think you need this here?

hwchase17 replied:
your right

jacoblee93 commented on libs/langchain/langchain/agents/xml/base.py:
```diff
@@ -97,3 +102,70 @@ async def aplan(
+
+
+def create_xml_agent(
+    llm: BaseLanguageModel, tools: Sequence[BaseTool], prompt: ChatPromptTemplate
```
No reason this one can't be `BasePromptTemplate`?

https://smith.langchain.com/hub/hwchase17/xml-agent-convo

hwchase17 replied:
your right

jacoblee93 commented on cookbook/human_approval.ipynb:
Do we need a redirect for these?

CC @baskaryan

jacoblee93 commented on docs/docs/modules/agents/agent_types/index.mdx:
```diff
@@ -1,52 +1,41 @@
-This agent uses the [ReAct](https://arxiv.org/pdf/2210.03629) framework to determine which tool to use
-based solely on the tool's description. Any number of tools can be provided.
-This agent requires that a description is provided for each tool.
+Whether this agent is intended for Chat Models (takes in messages, outputs message) or LLMs (takes in string, outputs string). The main thing this affects is the prompting strategy used. You can use an agent with a different type of model than it is intended for, but it likely won't product as good of results.
```
`product` -> `produce results of the same quality`

hwchase17 committed:
cr

jacoblee93 commented on docs/docs/modules/agents/quick_start.ipynb:
```diff
@@ -0,0 +1,683 @@
+   "source": [
+    "### Retriever\n",
+    "\n",
+    "We will also create a retriever over some data of our own. For a deeper explanation of each step here, see [this section](../../data_connection)"
```
Dead link, I think one too many `..`s

I've just been doing absolute path e.g.

`/docs/modules/data_connection/`

jacoblee93 commented on docs/docs/modules/agents/quick_start.ipynb:
```diff
@@ -0,0 +1,683 @@
+   "source": [
+    "## Run the agent\n",
+    "\n",
+    "We can now run the agent on a few queries! Note that for now, these are all **stateless** queries (it won't remember previous interactions."
```
Missing closing parenthesis

jacoblee93 commented on docs/docs/modules/agents/quick_start.ipynb:
```diff
@@ -0,0 +1,683 @@
+    {
+     "data": {
+      "text/plain": [
+       "{'input': 'how can langsmith help with testing?',\n",
```
Add a public LangSmith trace for folks to examine?

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite

hwchase17 committed:
cr

hwchase17 committed:
[core, langchain] modelio code improvements (#15277)

hwchase17 committed:
cr

hwchase17 committed:
[langchain] agents code changes (#15278)

<!-- Thank you for contributing to LangChain!

Please title your PR "<package>: <description>", where <package> is
whichever of langchain, community, core, experimental, etc. is being
modified.

Replace this entire comment with:
  - **Description:** a description of the change, 
  - **Issue:** the issue # it fixes if applicable,
  - **Dependencies:** any dependencies required for this change,
- **Twitter handle:** we announce bigger features on Twitter. If your PR
gets announced, and you'd like a mention, we'll gladly shout you out!

hwchase17 committed:
Merge branch 'master' into harrison/agents-rewrite
